lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -4298,7 +4298,11 @@ CheckedCastKind TypeChecker::typeCheckCheckedCast(Type fromType,
     //     print("Caught bar error")
     //   }
     // }
-    if (fromExistential) {
+    //
+    // Note: we relax the restriction if the type we're casting to is a class
+    // because it's possible that we might have a subclass that conforms to
+    // the protocol.
+    if (fromExistential && !toType->hasReferenceSemantics()) {
```

Comment:
I think the check should be for a conditional cast rather than the other type being a class. What do you think?

Reply:
But what happens if there's also `class Subclass: SR_6400_A, Error {}`?

Reply:
Yeah, I meant if you're doing a force cast then there would be a runtime crash if we get a `superclass` (and it doesn't conform to the protocol), where as a conditional cast would be fine (i.e. your code won't cause a crash). So, if you're force casting, you're accepting the risk that the code can crash, so we emit that diagnostic, but for conditional casts we don't, because that risk is handled.

Although I am fine with either approach (ignoring the check for non-final classes or ignoring the check for a conditional cast).